README.md
# vue debug helper for tampermonkey
Vue组件探测、统计、分析辅助脚本  
项目地址：[https://github.com/xxxily/vue-debug-helper](https://github.com/xxxily/vue-debug-helper)  
脚本安装地址：[https://greasyfork.org/scripts/444075](https://greasyfork.org/scripts/444075)

## 特性
* 支持生产环境下进行组件审查
* 按需打印组件的生命周期信息
* 支持全局组件注册情况观察
* 支持强制阻断组件的创建
* 支持当前激活的组件统计
* 支持对已销毁的组件进行统计
* 支持对对组件存活时间进行统计
* 支持进行组件模糊查找和根据uid精确查找
* 支持查找没有DOM关联的组件
* 支持给指定组件或全部组件注入空白数据
* 支持快速查看了解组件的关系链
* 支持实时打印性能观察(PerformanceObserver)结果
* 自动开启生产环境下的Vue调试模式

## 简介

Vue组件探测、统计、分析辅助脚本，主要用于分析当前应用的组件渲染情况和辅助检查组件的内存泄露情况，如果你是简单的vue应用，使用vue devtools 工具开发即可

## 其他说明
- 1、功能基于Vue.mixin，理论上只要支持全局mixin的vue版本都适用    
- 2、暂未对vue3进行测试，可能存在异常  
- 3、暂不打算对vue1进行兼容

## 快捷键列表
|  快捷键   | 说明    |
| --- | --- |
| shift+alt+i | 进入组件审查模式 |
| shift+alt+a | 全部组件混合统计 |
| shift+alt+l | 当前存活组件统计 |
| shift+alt+d | 已销毁组件统计 |
| shift+alt+c | 清空统计信息 |
| shift+alt+e | 数据注入（dd）或 取消数据注入（undd） |

![vue-debug-helper](./static/vue-debug-helper01.png)  

![vue-debug-helper](./static/vue-debug-helper02.png)

## 更新日志
* [https://github.com/xxxily/vue-debug-helper/blob/main/changeLog.md](https://github.com/xxxily/vue-debug-helper/blob/main/changeLog.md)

## TODO
* 区间统计功能
* 组件调用整体关系链输出
* 停用watch和computed
* 实现自定义快捷键
* 操作提示语之操作反馈优化
* 解决开启调试模式后，vue devtools 没正常开启问题
* 支持作为调试脚本，接入到应用
* vue3支持

changeLog.md
# 更新日志：
### 0.0.22 [2022/05/26]
* 优化对tampermonkey菜单的支持，实现动态注册菜单
* 解决审查组件时右键被抢注导致无法打开菜单的问题

### 0.0.21 [2022/05/23]
* 增加拦截接口请求功能
* 增加右键菜单精简模式
* 解决右键菜单样式受污染问题
* 允许缓存iframe下的接口

### 0.0.20 [2022/05/20]
* 实现对fetch发起的接口请求进行缓存的能力
* 组件审查时忽略transition组件，提升审查准确度

### 0.0.19 [2022/05/19]
* 右键菜单国际化改造、完善国际化信息
* 前置缓存接口功能，解决部分接口没法被缓存问题
* 修复组件审查功能组件名称错误问题
* 优化组件审查高亮框tip的展示逻辑

### 0.0.18 [2022/05/18]
* 组件审查高亮框增加组件名tips
* 优化样式设置方式，减少回流重绘

### 0.0.17 [2022/05/17]
* 优化大数据复制时信息没法正常打印问题
* 优化组件查找、阻断、生命周期信息打印的交互细节

### 0.0.16 [2022/05/16]
* 完善组件审查右键菜单功能
* 修复选择器测量无法被删除问题

### 0.0.15 [2022/05/15]
* 组件审查功能增加右键具体功能
* 解决关闭组件审查时右键没注销问题
* 修复接口缓存概要信息同步回收问题

### 0.0.14 [2022/05/14]
* 增加选择器之间的时间差测量工具
* 接口缓存优化对POST数据的支持 
* 修复接口缓存数据过期后没被移除的问题
* 修复关闭组件审查功能时组件标识没被移除的问题
* 增加右键相关基础库，推进inspect右键功能

### 0.0.13 [2022/05/13]
* 支持接口数据缓存能力
* 生命周期信息打印filters支持模糊匹配
* 组件查找filters支持模糊匹配

### 0.0.12 [2022/05/11]
* 增加打印性能观察结果的能力
* 优化过滤选项的内容提示

### 0.0.11 [2022/05/10]
* 增加元素审查功能（待完善）
* 生命周期信息增加组件路径信息
* 代码逻辑模块化划分调整

### 0.0.10 [2022/05/09]
* 修复aop导致的脚本不可用问题
* 优化组件阻断时频繁的错误消息

### 0.0.9 [2022/05/08]
* 实现组件完全阻断的能力
* 修复改写/还原Vue.component功能的提示错误
* 对重复注册的全局函数进行高亮提醒
* debug方法增加warn的支持

### 0.0.8 [2022/05/05]
* 增加改写Vue.component功能
* 增加全局组件查找功能
* 增加销毁组件总数统计功能

### 0.0.7 [2022/04/30]
* 增加开启或禁用vue-devtools功能
* 优化vue检测逻辑，提升检出效率

### 0.0.6 [2022/04/29]
* 增加搜索查找组建功能
* 增加阻断组件功能（实验）
* 实现配置项的记忆和同步
* 筛出所有没有dom关联的组件
* 统计信息输出展示方式优化
* 配置字段统一存放管理

### 0.0.5 [2022/04/28]
* 增加生命周期信息打印逻辑
* 增加自动开启Vue的调试模式
* 增加MutationObserver观察，加快Vue的检测
* 修复调用链信息不完整问题
* 优化Debug模块逻辑，减少不必要的字符串输出
* 优化创建时间信息输出

### 0.0.4 [2022/04/27]
* 国际化逻辑完善和功能调用提取
* 修复菜单注册错误问题

### 0.0.3 [2022/04/27]
* 调整数据注入逻辑
* 优化数据注入选项的输入体验
* 隐藏赞赏按钮

### 0.0.2 [2022/04/27]
* 补充功能描述

### 0.0.1 [2022/04/26]
* 发布首个版本

src/vueDebugger/inspect.js
/*!
 * @name         inspect.js
 * @description  vue组件审查模块
 * @version      0.0.1
 * @author       xxxily
 * @date         2022/05/10 18:25
 * @github       https://github.com/xxxily
 */
import helper from './helper'
import debug from './debug'
import i18n from './i18n'
import functionCall from './functionCall'
import {
  openInTab,
  copyToClipboard
} from './utils'

const overlaySelector = 'vue-debugger-overlay'
const $ = window.$
let currentComponent = null

const inspect = {
  findComponentsByElement (el) {
    let result = null
    let deep = 0
    let parent = el
    while (parent) {
      if (deep >= 50) {
        break
      }

      if (parent.__vue__) {
        result = parent
        break
      }

      deep++
      parent = parent.parentNode
    }

    return result
  },

  getComponentInstance (el) {
    let vueComponent = el && el.__vue__ ? el.__vue__ : null

    /* 忽略transition */
    if (vueComponent && vueComponent?.$options._componentTag === 'transition' && vueComponent.$parent) {
      vueComponent = vueComponent.$parent
    }

    return vueComponent
  },

  initContextMenu () {
    if (this._hasInitContextMenu_) {
      return
    }

    function createComponentMenuItem (vueComponent, deep = 0) {
      let componentMenu = {}
      if (vueComponent) {
        helper.methods.initComponentInfo(vueComponent)

        componentMenu = {
          consoleComponent: {
            name: `${i18n.t('contextMenu.consoleComponent')} <${vueComponent._componentName}>`,
            icon: 'fa-eye',
            callback: function (key, options) {
              debug.log(`[vueComponent] ${vueComponent._componentTag}`, vueComponent)
            }
          },
          consoleComponentData: {
            name: `${i18n.t('contextMenu.consoleComponentData')} <${vueComponent._componentName}>`,
            icon: 'fa-eye',
            callback: function (key, options) {
              debug.log(`[vueComponentData] ${vueComponent._componentTag}`, vueComponent.$data)
            }
          },
          consoleComponentProps: {
            name: `${i18n.t('contextMenu.consoleComponentProps')} <${vueComponent._componentName}>`,
            icon: 'fa-eye',
            callback: function (key, options) {
              debug.log(`[vueComponentProps] ${vueComponent._componentTag}`, vueComponent.$props)
            }
          }
          // consoleComponentChain: {
          //   name: `${i18n.t('contextMenu.consoleComponentChain')} <${vueComponent._componentName}>`,
          //   icon: 'fa-eye',
          //   callback: function (key, options) {
          //     debug.log(`[vueComponentMethods] ${vueComponent._componentTag}`, vueComponent._componentChain)
          //   }
          // }
        }
      }

      if (vueComponent.$parent && deep <= 5) {
        componentMenu.parentComponent = {
          name: `${i18n.t('contextMenu.consoleParentComponent')} <${vueComponent.$parent._componentName}>`,
          icon: 'fa-eye',
          items: createComponentMenuItem(vueComponent.$parent, deep + 1)
        }
      }

      const file = vueComponent.options?.__file || vueComponent.$options?.__file || ''
      let copyFilePath = {}
      if (file) {
        copyFilePath = {
          copyFilePath: {
            name: `${i18n.t('contextMenu.copyFilePath')}`,
            icon: 'fa-copy',
            callback: function (key, options) {
              debug.log(`[componentFilePath ${vueComponent._componentName}] ${file}`)
              copyToClipboard(file)
            }
          }
        }
      }

      componentMenu.componentAction = {
        name: `${i18n.t('contextMenu.componentAction')} <${vueComponent._componentName}>`,
        icon: 'fa-cog',
        items: {
          ...copyFilePath,
          copyComponentName: {
            name: `${i18n.t('contextMenu.copyComponentName')} <${vueComponent._componentName}>`,
            icon: 'fa-copy',
            callback: function (key, options) {
              copyToClipboard(vueComponent._componentName)
            }
          },
          copyComponentData: {
            name: `${i18n.t('contextMenu.copyComponentData')} <${vueComponent._componentName}>`,
            icon: 'fa-copy',
            callback: function (key, options) {
              const data = JSON.stringify(vueComponent.$data, null, 2)
              debug.log(`[vueComponentData] ${vueComponent._componentName}`, JSON.parse(data))
              debug.log(data)
              copyToClipboard(data)
            }
          },
          copyComponentProps: {
            name: `${i18n.t('contextMenu.copyComponentProps')} <${vueComponent._componentName}>`,
            icon: 'fa-copy',
            callback: function (key, options) {
              const props = JSON.stringify(vueComponent.$props, null, 2)
              debug.log(`[vueComponentProps] ${vueComponent._componentName}`, JSON.parse(props))
              debug.log(props)
              copyToClipboard(props)
            }
          },
          // copyComponentTag: {
          //   name: `${i18n.t('contextMenu.copyComponentTag')} <${vueComponent._componentName}>`,
          //   icon: 'fa-copy',
          //   callback: function (key, options) {
          //     copyToClipboard(vueComponent._componentTag)
          //   }
          // },
          copyComponentUid: {
            name: `${i18n.t('contextMenu.copyComponentUid')} -> ${vueComponent._uid}`,
            icon: 'fa-copy',
            callback: function (key, options) {
              copyToClipboard(vueComponent._uid)
            }
          },
          copyComponentChian: {
            name: `${i18n.t('contextMenu.copyComponentChian')}`,
            icon: 'fa-copy',
            callback: function (key, options) {
              debug.log(`[vueComponentChain] ${vueComponent._componentName}`, vueComponent._componentChain)
              copyToClipboard(vueComponent._componentChain)
            }
          },
          findComponents: {
            name: `${i18n.t('contextMenu.findComponents')} <${vueComponent._componentName}>`,
            icon: 'fa-search',
            callback: function (key, options) {
              functionCall.findComponents(vueComponent._componentName)
            }
          },
          printLifeCycleInfo: {
            name: `${i18n.t('contextMenu.printLifeCycleInfo')} <${vueComponent._componentName}>`,
            icon: 'fa-print',
            callback: function (key, options) {
              functionCall.printLifeCycleInfo(vueComponent._componentName)
            }
          },
          blockComponents: {
            name: `${i18n.t('contextMenu.blockComponents')} <${vueComponent._componentName}>`,
            icon: 'fa-ban',
            callback: function (key, options) {
              functionCall.blockComponents(vueComponent._componentName)
            }
          }
        }
      }

      return componentMenu
    }

    $.contextMenu({
      selector: 'body.vue-debug-helper-inspect-mode',
      zIndex: 2147483647,
      className: 'vue-debug-helper-context-menu',
      build: function ($trigger, e) {
        const conf = helper.config
        const vueComponent = inspect.getComponentInstance(currentComponent)

        let componentMenu = {}
        if (vueComponent) {
          componentMenu = createComponentMenuItem(vueComponent)
          componentMenu.componentMenuSeparator = '---------'
        }

        const componentsStatisticsInfo = {
          componentsStatistics: {
            name: i18n.t('debugHelper.componentsStatistics'),
            icon: 'fa-thin fa-info-circle',
            callback: functionCall.componentsStatistics
          },
          componentsSummaryStatisticsSort: {
            name: i18n.t('debugHelper.componentsSummaryStatisticsSort'),
            icon: 'fa-thin fa-info-circle',
            callback: functionCall.componentsSummaryStatisticsSort
          },
          destroyStatisticsSort: {
            name: i18n.t('debugHelper.destroyStatisticsSort'),
            icon: 'fa-regular fa-trash',
            callback: functionCall.destroyStatisticsSort
          },
          clearAll: {
            name: i18n.t('debugHelper.clearAll'),
            icon: 'fa-regular fa-close',
            callback: functionCall.clearAll
          }
        }

        const commonMenu = {
          findComponents: {
            name: i18n.t('debugHelper.findComponents'),
            icon: 'fa-regular fa-search',
            callback: () => {
              functionCall.findComponents()
            }
          },
          blockComponents: {
            name: i18n.t('debugHelper.blockComponents'),
            icon: 'fa-regular fa-ban',
            callback: () => {
              functionCall.blockComponents()
            }
          },
          printLifeCycleInfo: {
            name: conf.lifecycle.show ? i18n.t('debugHelper.notPrintLifeCycleInfo') : i18n.t('debugHelper.printLifeCycleInfo'),
            icon: 'fa-regular fa-life-ring',
            callback: () => {
              conf.lifecycle.show ? functionCall.notPrintLifeCycleInfo() : functionCall.printLifeCycleInfo()
            }
          },
          dd: {
            name: conf.dd.enabled ? i18n.t('debugHelper.undd') : i18n.t('debugHelper.dd'),
            icon: 'fa-regular fa-arrows-alt',
            callback: conf.dd.enabled ? functionCall.undd : functionCall.dd
          },
          toggleHackVueComponent: {
            name: conf.hackVueComponent ? i18n.t('debugHelper.hackVueComponent.unhack') : i18n.t('debugHelper.hackVueComponent.hack'),
            icon: 'fa-regular fa-bug',
            callback: functionCall.toggleHackVueComponent
          },
          componentFunSeparator: '---------',
          toggleAjaxCache: {
            name: conf.ajaxCache.enabled ? i18n.t('debugHelper.ajaxCacheStatus.off') : i18n.t('debugHelper.ajaxCacheStatus.on'),
            icon: 'fa-regular fa-database',
            callback: functionCall.toggleAjaxCache
          },
          clearAjaxCache: {
            name: i18n.t('debugHelper.clearAjaxCache'),
            icon: 'fa-regular fa-database',
            callback: functionCall.clearAjaxCache
          },
          toggleBlockAjax: {
            name: conf.blockAjax.enabled ? i18n.t('debugHelper.blockAjax.disable') : i18n.t('debugHelper.blockAjax.enabled'),
            icon: 'fa-regular fa-ban',
            callback: functionCall.toggleBlockAjax
          },
          togglePerformanceObserver: {
            name: conf.performanceObserver.enabled ? i18n.t('debugHelper.performanceObserverStatus.off') : i18n.t('debugHelper.performanceObserverStatus.on'),
            icon: 'fa-regular fa-paint-brush',
            callback: functionCall.togglePerformanceObserver
          },
          measureSelectorInterval: {
            name: i18n.t('debugHelper.measureSelectorInterval'),
            icon: 'fa-regular fa-clock-o',
            callback: functionCall.measureSelectorInterval
          },
          commonEndSeparator: '---------'
        }

        const moreMenu = {
          ...(conf.contextMenu.simplify ? commonMenu : {}),
          help: {
            name: i18n.t('help'),
            icon: 'fa-regular fa-question-circle',
            items: {
              docs: {
                name: i18n.t('docs'),
                icon: 'fa-regular fa-book',
                callback: () => { openInTab('https://github.com/xxxily/vue-debug-helper') }
              },
              about: {
                name: i18n.t('about'),
                icon: 'fa-regular fa-info-circle',
                callback: () => { openInTab('https://github.com/xxxily/vue-debug-helper') }
              },
              issues: {
                name: i18n.t('issues'),
                icon: 'fa-regular fa-bug',
                callback: () => { openInTab('https://github.com/xxxily/vue-debug-helper/issues') }
              },
              update: {
                name: i18n.t('update'),
                icon: 'fa-regular fa-refresh',
                callback: () => { openInTab('https://greasyfork.org/zh-CN/scripts/444075') }
              }
            }
          },
          toggleSimplifyMode: {
            name: conf.contextMenu.simplify ? i18n.t('debugHelper.simplifyMode.disable') : i18n.t('debugHelper.simplifyMode.enabled'),
            icon: 'fa-regular fa-compress',
            callback: functionCall.toggleSimplifyMode
          },
          toggleInspect: {
            name: conf.inspect.enabled ? i18n.t('debugHelper.inspectStatus.off') : i18n.t('debugHelper.inspectStatus.on'),
            icon: 'fa-regular fa-eye',
            callback: functionCall.toggleInspect
          }
        }

        const menu = {
          callback: function (key, options) {
            debug.log(`[contextMenu] ${key}`)
          },
          items: {
            refresh: {
              name: i18n.t('refreshPage'),
              icon: 'fa-refresh',
              callback: function (key, options) {
                window.location.reload()
              }
            },
            sep0: '---------',
            ...componentMenu,
            ...componentsStatisticsInfo,
            statisticsSeparator: '---------',
            ...(conf.contextMenu.simplify ? {} : commonMenu),
            more: {
              name: i18n.t('contextMenu.more'),
              icon: 'fa-ellipsis-h',
              items: {
                ...moreMenu
              }
            },
            quit: {
              name: i18n.t('quit'),
              icon: 'fa-close',
              callback: function ($element, key, item) {
                return 'context-menu-icon context-menu-icon-quit'
              }
            }
          }
        }

        return menu
      }
    })

    /* 主动触发右键，解决部分右键菜单已被提前注册且禁止冒泡导致右键无法触发的问题 */
    document.addEventListener('contextmenu', function (e) {
      helper.config.inspect.enabled && $('body.vue-debug-helper-inspect-mode').contextMenu({ x: e.pageX, y: e.pageY })
    }, true)

    this._hasInitContextMenu_ = true
  },

  setOverlay (el) {
    let overlay = document.querySelector('#' + overlaySelector)
    if (!overlay) {
      overlay = document.createElement('div')
      overlay.id = overlaySelector

      const infoBox = document.createElement('div')
      infoBox.className = 'vue-debugger-component-info'

      const styleDom = document.createElement('style')
      styleDom.appendChild(document.createTextNode(`
        .vue-debug-helper-context-menu {
          font-size: 14px;
        }
        #${overlaySelector} {
          position: fixed;
          z-index: 2147483647;
          background-color: rgba(65, 184, 131, 0.15);
          padding: 5px;
          font-size: 11px;
          pointer-events: none;
          box-size: border-box;
          border-radius: 3px;
          overflow: visible;
        }

        #${overlaySelector} .vue-debugger-component-info {
          position: absolute;
          top: -30px;
          left: 0;
          line-height: 1.5;
          display: inline-block;
          padding: 4px 8px;
          border-radius: 3px;
          background-color: #fff;
          font-family: monospace; 
          font-size: 11px;
          color: rgb(51, 51, 51); 
          text-align: center; 
          border: 1px solid rgba(65, 184, 131, 0.5); 
          background-clip: padding-box;
          pointer-events: none;
          white-space: nowrap;
        }
      `))

      overlay.appendChild(infoBox)
      document.body.appendChild(styleDom)
      document.body.appendChild(overlay)
    }

    /* 批量设置样式，减少样式扰动 */
    const rect = el.getBoundingClientRect()
    const overlayStyle = [
      `width: ${rect.width}px;`,
      `height: ${rect.height}px;`,
      `top: ${rect.top}px;`,
      `left: ${rect.left}px;`,
      'display: block;'
    ].join(' ')
    overlay.setAttribute('style', overlayStyle)

    const vm = inspect.getComponentInstance(el)
    if (vm) {
      helper.methods.initComponentInfo(vm)
      const name = vm._componentName || vm._componentTag || vm._uid
      const infoBox = overlay.querySelector('.vue-debugger-component-info')

      infoBox.innerHTML = [
        '<span style="opacity: 0.6;">&lt;</span>',
        `<span style="font-weight: bold; color: rgb(9, 171, 86);">${name}</span>`,
        '<span style="opacity: 0.6;">&gt;</span>',
        `<span style="opacity: 0.5; margin-left: 6px;">${Math.round(rect.width)}<span style="margin-right: 2px; margin-left: 2px;">×</span>${Math.round(rect.height)}</span>`
      ].join('')

      rect.y < 32 ? (infoBox.style.top = '0') : (infoBox.style.top = '-30px')
    }

    $(document.body).addClass('vue-debug-helper-inspect-mode')
    inspect.initContextMenu()
  },

  clearOverlay () {
    $(document.body).removeClass('vue-debug-helper-inspect-mode')
    const overlay = document.querySelector('#vue-debugger-overlay')
    if (overlay) {
      overlay.style.display = 'none'
    }
  },

  init (Vue) {
    document.body.addEventListener('mouseover', (event) => {
      if (!helper.config.inspect.enabled) {
        return
      }

      const componentEl = inspect.findComponentsByElement(event.target)

      if (componentEl) {
        currentComponent = componentEl
        inspect.setOverlay(componentEl)
      } else {
        currentComponent = null
      }
    })
  }
}

export default inspect
